src/solvers/refinement/string_refinement.cpp

```diff
@@ -199,321 +217,304 @@ static std::vector<exprt> generate_instantiations(
+        it.next_sibling_or_parent();
+      }
+      else
+      {
+        symbol_exprt length_sym=
+            generator.fresh_symbol("data_length", java_int_type());
+        symbol_exprt array_sym=generator.fresh_symbol(
+              "data_array", array_typet(java_char_type(), length_sym));
+
+        array_of_pointers.insert(std::make_pair(*it, array_sym));
+        it.next_sibling_or_parent();
+      }
     }
-  }
-  for(const auto &it : created_strings)
-  {
-    if(const auto str=expr_cast<string_exprt>(it))
-    {
-      exprt length=get(str->length());
-      exprt content=str->content();
-      replace_expr(symbol_resolve, content);
-      found_length[content]=length;
-     }
+    else
+      ++it;
   }
 }
 
-/// add lemmas representing the setting of an expression to a given value
-/// \par parameters: an expression and the value to set it to
+void replace_symbols_in_equations(
+  const union_find_replacet &solver,
+  std::vector<equal_exprt> &equations)
+{
+  for(equal_exprt &eq : equations)
+    solver.replace_expr(eq);
+}
+
+/// Add equation to `m_equation_list` or give them to `supert::set_to`
+/// \param expr: an expression
+/// \param value: the value to set it to
 void string_refinementt::set_to(const exprt &expr, bool value)
 {
   PRECONDITION(expr.type().id()==ID_bool);
   PRECONDITION(equality_propagation);
 
-  if(expr.id()==ID_equal)
+  if(expr.id()==ID_equal && value)
   {
-    equal_exprt eq_expr=to_equal_expr(expr);
-    const exprt &lhs=eq_expr.lhs();
-    const exprt &rhs=eq_expr.rhs();
-
-    // The assignment of a string equality to false is not supported.
-    PRECONDITION(value || !is_char_array(ns, rhs.type()));
-    PRECONDITION(value || !is_refined_string_type(rhs.type()));
-
-    PRECONDITION(lhs.id()==ID_symbol || !is_char_array(ns, rhs.type()));
-    PRECONDITION(lhs.id()==ID_symbol || !is_refined_string_type(rhs.type()));
+    const equal_exprt &eq_expr=to_equal_expr(expr);
+    equations.push_back(eq_expr);
+  }
+  else
+  {
+    if(has_char_array_subexpr(expr, ns))
+      warning() << "WARNING: string_refinement.cpp: "
+                   "non string equation has char array subexpr";
+    supert::set_to(expr, value);
+  }
+}
 
-    // If lhs is not a symbol, let supert::set_to() handle it.
+/// Add association for each char pointer in the equation
+static union_find_replacet symbol_solver_from_equations(
+  const std::vector<equal_exprt> &equations,
+  const namespacet &ns,
+  messaget::mstreamt &stream)
+{
+  const auto eom=messaget::eom;
+  const std::string log_message=
+    "WARNING string_refinement.cpp symbol_solver_from_equations:";
+  union_find_replacet solver;
+  for(const equal_exprt &eq : equations)
+  {
+    const exprt &lhs=eq.lhs();
+    const exprt &rhs=eq.rhs();
     if(lhs.id()!=ID_symbol)
     {
-      non_string_axioms.emplace_back(expr, value);
-      return;
+      stream << log_message << "non symbol lhs: " << from_expr(ns, "", lhs)
```

Comment:
actually sometimes a string pointer is allocated a void* pointer which does not seem to matter for the solver, so it could have a char pointer subtype. I'm not sure of what check we should have here.